### DISCUSSION TITLE
Make VALI more object oriented

### DISCUSSION THREAD

**Discussion by kqv1** [RomanArzumanyan/VALI#99] [Ideas]
While the current library's working well, @RomanArzumanyan may consider making it more object oriented,

I would suggest to move the **Converter**, **Resizer**, and **Downloader** classes to method inside **Surface** / **Frame**.
**PySurfaceConverter.Run** => **Surface.Convert**
**PySurfaceResizer.Run** => **Surface.Resize**
**PySurfaceDownloader.Run** => **Surface.Down**

Similarly, add a new **Frame** class with **Frame.Convert** and **Frame.Upload**

**Comment by RomanArzumanyan** [comment]
Hi @kqv1 
Thank you for bringing this up!

First of all, I would like to take a liberty and put on my grumpy sw eng hat.
IIRC, OOP is about 3 major ideas, which are
- Inheritance
- Encapsulation
- Polymorphism

Although renaming class methods may be a good idea, I can't really see how does it make code more object oriented ?

WRT to introduction of `Frame` class, I can't agree with that. VPF and VALI had been following a naming convention that picture in RAM is called `Frame` while the same thing in vRAM is called `Surface`. Hence you can tell what does the class do by it's name:
- `PySurfaceResizer` resizes pictures in vRAM
- `PyFrameConverter` converts pictures in RAM between different formats
- and so on

I'd like not to add `Frame` class because VALI relies on numpy ndarray instead.

However, I can imagine a situation when e. g. `Resize` or `Convert` transformations are static methods of some other class. Something like that:
```python
fhd_frame = np.ndarray(shape=(1080, 1920), dtype=np.uint8)
hd_frame = np.ndarray(shape=(720, 1280), dtype=np.uint8)
vali.Transform.CPU.Resize(fhd_frame, hd_frame)

surf_rgb = vali.Surface.Make(vali.PixelFormat.RGB, 1920, 1080)
surf_yuv = vali.Surface.Make(vali.PixelFormat.YUV420, 1920, 1080)
vali.Transform.CUDA.Convert(surf_yuv, surf_rgb)
```

But there's a caveat here. VALI classes aren't thread safe. If you like to span multiple threads, you may create a `PySurfaceConverter` class in each and every thread. But if you implement those transformation as static methods or just functions within `python_vali` module, you shall get rid of all internal state or add sync primitives. 
Otherwise, your multi threaded `Surface` conversion will become a mess for certain formats that rely on interim scratch buffers.

`Upload` and `Download` are a bit different.
I was thinking for a while about adding the [Cuda Array Interface](https://numba.readthedocs.io/en/stable/cuda/cuda_array_interface.html)  support (similar to DLPack). That would be the direction I'd like to move in. Completely get rid of `Upload` and `Download` for good.

**Reply by kqv1** [reply]
I just suggest keeping actions `Convert`, `Resize`, `Download`, `Upload ` as methods to their object (`Surface` and `Frame`). Now they are independent classes with one `Run `method, which is not very intuitive to me (a new developer who learning to use VALI).

Thus, I think it's not violated the principle
> VPF and VALI had been following a naming convention that picture in RAM is called `Frame` while the same thing in vRAM is called `Surface`. Hence you can tell what does the class do by it's name:
> 
> * `PySurfaceResizer` resizes pictures in vRAM
> * `PyFrameConverter` converts pictures in RAM between different formats
> * and so on

My suggestions are from the point of someone who use the library (not someone who created it). Thus, I maynot understand reasons behind.

**Reply by RomanArzumanyan** [reply]
> My suggestions are from the point of someone who use the library (not someone who created it)

Yeah, I understand. Users PoV is much appreciated !

> I just suggest keeping actions Convert, Resize, Download, Upload  as methods to their object (Surface and Frame).

The point is, there's no such object as `Frame`. If you make one up, it will be a collection of static methods that do things with numpy arrays:

```python
fhd_frame = np.ndarray(shape=(1080, 1920), dtype=np.uint8)
hd_frame = np.ndarray(shape=(720, 1280), dtype=np.uint8)
vali.Frame.Resize(fhd_frame, hd_frame)
```

Unlike that, `Surface` is actual class which may own the memory. So the interface will be different:
```python
surf_rgb = vali.Surface.Make(vali.PixelFormat.RGB, 1920, 1080)
surf_yuv = vali.Surface.Make(vali.PixelFormat.YUV420, 1920, 1080)
surf_yuv.Convert(surf_rgb)
```

Doesn't look attractive IMO.
Maybe I'm missing something ?

**Reply by kqv1** [reply]
for the `Surface` syntax above, it looks cleaner to me (more intuitive than `PySurfaceConverter.Run `)

for the `Frame` syntax, can we create a simple class that wrap `np.ndarray`? And keep the action `Resize`/`Upload` as instance methods.
```python
fhd_frame = vali.Frame(1080, 1920)
hd_frame = vali.Frame(720, 1280)
fhd_frame.Resize(hd_frame)
```

**Reply by RomanArzumanyan** [reply]
I'm not going to do this because it's a huge risk for marginal gain.
Here are couple reasons that come from the top of my head:

- Problem of inheritance: `py::array` and `py::array_t` on cpp side, `numpy.ndarray` on python side. I have no idea how to do this. Pybind11 does the casting for me, will it do the same for a child class ?
- Yet another data class specific to a small project nobody wants to get into.

All in all, I'd like to take an opposite way and support more standards de-facto like DLPack and CAI.

**Reply by kqv1** [reply]
Thanks, I understand your view.

**Comment by RomanArzumanyan** [comment]
@kqv1 

WRT your comment https://github.com/RomanArzumanyan/VALI/pull/98#discussion_r1743833964
Yes, all VALI classes operate in-place for a very simple reason: performance.

```python
# CUDA 2D pitched memory alloc per every call
surf_nv12 = nv_dec.DecodeSingleSurface()

# Allocate just once, keep all your Surfaces somewhere in memory pool
# Steal torch cuda tensor memory via DLPack, go wild !
surf_nv12 = vali.Surface.make(...)
while True:
    success, info = nv_dec.DecodeSingleSurface(surf_nv12)
    if not success:
        print(info)
        break
```

**Reply by kqv1** [reply]
Totally agree, performance should be the highest priority
